src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -2,32 +2,124 @@
+        this.prev = null;
+        this.next = null;
+    }
+}
+
+
+class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node head;
+    private Node tail;
+    private int size;
+
+    public MyLinkedList() {
+        head = null;
+        tail = null;
+        size = 0;
+    }
+
     @Override
-    public void add(T value) {
+    public int size() {
+        return size;
     }
 
     @Override
-    public void add(T value, int index) {
+    public boolean isEmpty() {
+        return size == 0;
     }
 
     @Override
-    public void addAll(List<T> list) {
+    public void add(T value) {
+        Node<T> newNode = new Node<>(value);
+
+        if (isEmpty()) {
+            head = newNode;
+            tail = newNode;
+        } else {
+            tail.next = newNode;
+            newNode.prev = tail;
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
-    public T get(int index) {
-        return null;
+    public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException(
+                    "Index outside the list");
+        }
+        Node<T> newNode = new Node<>(value);
+
+        if (index == 0) {
+            if (isEmpty()) {
+                head = newNode;
+                tail = newNode;
+            } else {
+                newNode.next = head;
+                head.prev = newNode;
+                head = newNode;
+            }
+        } else if (index == size) {
+            add(value);
+            return;
+        } else {
+            Node<T> current = head;
+            for (int i = 0; i < index; i++) {
+                current = current.next;
+            }
+            newNode.next = current;
+            newNode.prev = current.prev;
+            current.prev.next = newNode;
+            current.prev = newNode;
+        }
+        size++;
     }
 
     @Override
-    public T set(T value, int index) {
-        return null;
+    public void addAll(List<T> list) {
+
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException(
+                    "Index outside the list");
+        }
+
+        Node<T> toRemove;
+        if (index == 0) {
+            toRemove = head;
+            if (head == tail) {
+                head = null;
+                tail = null;
+            } else {
+                head = head.next;
+                head.prev = null;
+            }
+        } else if (index == size - 1) {
+            toRemove = tail;
+            tail = tail.prev;
+            tail.next = null;
+        } else {
+            toRemove = head;
+            for (int i = 0; i < index; i++) {
+                toRemove = toRemove.next;
+            }
+            toRemove.prev.next = toRemove.next;
+            toRemove.next.prev = toRemove.prev;
+        }
+        size--;
+        return toRemove.data;
```

Comment:
Create a separate method `unlink(Node<T> node)` to encapsulate the logic of unlinking a node from the list. This will avoid duplication in methods where a node needs to be removed.